src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PreReleaseUserService.cs

```diff
@@ -231,130 +234,185 @@ public class PreReleaseUserService : IPreReleaseUserService
-                )
-                .AnyAsync();
-
-            if (hasExistingReleaseInvite)
+            return await CreateAcceptedUserReleaseInvite(release, email, user)
+                .OnSuccess(_ => new PreReleaseUserViewModel(email));
+        }
+
+        private async Task<Either<ActionResult, Unit>> CreateUserReleaseInvite(Release release, string email)
+        {
+            if (!await UserHasPreReleaseInvite(release.Id, email))
             {
-                return ValidationActionResult(UserAlreadyExists);
+                var sendEmail = release.ApprovalStatus == ReleaseApprovalStatus.Approved;
+                await _context.AddAsync(
+                    new UserReleaseInvite
+                    {
+                        Email = email.ToLower(),
+                        ReleaseId = release.Id,
+                        Role = ReleaseRole.PrereleaseViewer,
+                        EmailSent = sendEmail,
+                        Created = UtcNow,
+                        CreatedById = _userService.GetUserId()
+                    }
+                );
+                await _context.SaveChangesAsync();
+
+                if (sendEmail)
+                {
+                    return await SendPreReleaseInviteEmail(release, email, isNewUser: true);
+                }
             }
 
             return Unit.Instance;
         }
 
-        private async Task CreateNewUserReleaseInvite(Guid releaseId, string email, bool isExistingUser, bool emailAlreadySent)
+        private async Task<Either<ActionResult, Unit>> CreateAcceptedUserReleaseInvite(Release release, string email,
+            User user)
         {
-            var hasExistingInvite = await _context
-                .UserReleaseInvites
-                .AsQueryable()
-                .Where(
-                    i =>
-                        i.ReleaseId == releaseId
-                        && i.Email.ToLower() == email.ToLower()
-                        && i.Role == ReleaseRole.PrereleaseViewer
-                )
-                .AnyAsync();
-
-            if (hasExistingInvite)
+            if (!await UserHasPreReleaseInvite(release.Id, email))
             {
-                return;
-            }
+                var sendEmail = release.ApprovalStatus == ReleaseApprovalStatus.Approved;
+
+                // Create a new release invite
+                // The invite is already accepted if the user exists
+                await _context.AddAsync(
+                    new UserReleaseInvite
+                    {
+                        Email = email.ToLower(),
+                        ReleaseId = release.Id,
+                        Role = ReleaseRole.PrereleaseViewer,
+                        Accepted = true,
+                        EmailSent = sendEmail,
+                        Created = UtcNow,
+                        CreatedById = _userService.GetUserId()
+                    }
+                );
```

Comment:
Similar to above - shouldn't we wait for the result of `SendPreReleaseInviteEmail` before saving the invite into the database as sent?

Reply:
Good point, thanks. Altered this so that it doesn't get saved if the email fails to send.